File: framework/java/src/main/java/soda/unittest/work/parse/NestedIntegerConverter.java
package soda.unittest.work.parse;

import soda.leetcode.DefaultNestedInteger;
import soda.leetcode.NestedInteger;

import java.util.ArrayList;
import java.util.List;

public class NestedIntegerConverter implements ObjectConverter<NestedInteger, Object> {
    @Override
    public NestedInteger fromJsonSerializable(Object o) {
        return parse(o);
    }

    @Override
    public Object toJsonSerializable(NestedInteger nestedInteger) {
        return serialize(nestedInteger);
    }

    @SuppressWarnings("unchecked")
    public static NestedInteger parse(Object obj) {
        if (obj instanceof Integer) {
            return new DefaultNestedInteger((Integer) obj);
        }
        var ni = new DefaultNestedInteger();
        for (var a : (List<Object>) obj) {
            ni.add(parse(a));
        }
        return ni;
    }

    public static Object serialize(NestedInteger ni) {
        if (ni.isInteger()) {
            return ni.getInteger();
        }
        var list = new ArrayList<Object>();
        for (var e : ni.getList()) {
            list.add(serialize(e));
        }
        return list;
    }
}

File: framework/csharp/soda/unittest/conv/NestedIntegerConverter.cs
using Newtonsoft.Json.Linq;
using Soda.Leetcode;

namespace Soda.Unittest.Conv;

public static class NestedIntegerConverter
{
    public static ObjectConverter<NestedInteger> Create()
    {
        return new ObjectConverter<NestedInteger>(parse, serialize);
    }

    public static ObjectConverter<IList<NestedInteger>> CreateForList()
    {
        return new ObjectConverter<IList<NestedInteger>>(
            (JToken jt) => jt.Select(e => parse(e)).ToList(),
            (IList<NestedInteger> ns) => JToken.FromObject(ns.Select(e => serialize(e)))
        );
    }

    private static NestedInteger parse(JToken jt)
    {
        if (jt.Type == JTokenType.Integer) {
            return new NestedIntegerImpl(jt.ToObject<int>());
        }
        var ni = new NestedIntegerImpl();
        foreach (var a in jt) {
            ni.Add(parse(a));
        }
        return ni;
    }

    private static JToken serialize(NestedInteger ni)
    {
        if (ni.IsInteger()) {
            return JToken.FromObject(ni.GetInteger());
        }
        var list = new JArray();
        foreach (var e in ni.GetList()) {
            list.Add(serialize(e));
        }
        return list;
    }
}